[MPDX-8227] Show multiple 14 month partner currency tables

canac opened the pull request:
## Description

The 14 month partner currency report was only showing one table even when there were multiple currencies given. This PR fixes that in the UI and in the CSV export. It also does a fair amount of refactoring to move complex logic into independent helper functions. Lastly, it adds tests for those helper functions.

@dr-bizz The print styles you added were causing weird issues with multiple tables, so I had to remove a lot of them. It still seems to print well, but I'd love your feedback on whether I unintentionally removed some crucial print styles.

@wrandall22 This should fix the rest of the inconsistencies I was seeing in CSV exports. There are still differences in rounding (Angular appears to round down, and React rounds to the nearest dollar, which I believe is more correct). I'm investigating one more discrepancy involving totals, but Angular seems clearly wrong in this case because the export includes a number that's no where on the table in the UI.

[MPDX-8227](https://jira.cru.org/browse/MPDX-8227)

## Checklist:

- [x] I have given my PR a title with the format "MPDX-(JIRA#) (summary sentence max 80 chars)"
- [x] I have applied the appropriate labels. (_Add the label "On Staging" to get the branch automatically merged into staging._)
- [x] I have requested a review from another person on the project

wrandall22 commented:
Some things I still see as discrepancies (looking at partner currency export):

React has (and Angular does not have):
Commitment amount, Commitment currency, Commitment frequency (these are all in Angular, but formatted differently)
Committed monthly equivalent, In hand monthly equivalent, Missing in hand monthly equivalent, In hand special gifts, In hand date range

Angular has (and React does not have):
Pledge (this is a combined field of those mentioned above), Average, Minimum, Maximum

Angular formats the status in a more readable manner. The dates are formatted differently between the two exports. Angular has the currency symbol, React does not.

wrandall22 commented:
The multiple tables stuff is looking pretty good from my perspective. I do notice that the Avg and Min now get calculated differently between Angular and React on both reports (Angular has 42, React has 84). React is probably more accurate in this case.

canac commented:
> React has (and Angular does not have): Commitment amount, Commitment currency, Commitment frequency (these are all in Angular, but formatted differently) Committed monthly equivalent, In hand monthly equivalent, Missing in hand monthly equivalent, In hand special gifts, In hand date range
>
> Angular has (and React does not have): Pledge (this is a combined field of those mentioned above), Average, Minimum, Maximum

This is because [the PR](https://github.com/CruGlobal/mpdx_web/pull/2243/files) to add those fields to Angular is in staging but never merged into production. If you compare an export from this branch's preview environment and an export from Angular staging, they're nearly identical, ignoring rounding differences.

> Angular formats the status in a more readable manner. The dates are formatted differently between the two exports. Angular has the currency symbol, React does not.

I changed the date formatting to be more similar (i.e. MM YY for en-US). It's not always identical because we now format dates in a locale-dependent way. I added the currency symbol to React as well.

Angular also gives a different value for the totals of foreign currencies. Here is a condensed version of the output with extraneous totals removed. The totals for GPB and UGS should be 10,000 and 333,333,333, respectively, not 12,719 and 895. I believe that the React is correct in this case.

# Angular

| Currency                | GBP               | £                 |                     |                      |           |        |                                        |
| ----------------------- | ----------------- | ----------------- | ------------------- | -------------------- | --------- | ------ | -------------------------------------- |
| Partner                 | Status            | Commitment Amount | Commitment Currency | Commitment Frequency | 24-Sep    | 24-Aug | Total (last month excluded from total) |
| Duck, Daffy and Daphney | partner_financial | 30                | USD                 | Monthly              | 0         | 10000  | 10000                                  |
| Totals                  |                   |                   |                     |                      | 0         | 10000  | 12719                                  |
| Currency                | UGS               | UGS               |                     |                      |           |        |                                        |
| Partner                 | Status            | Commitment Amount | Commitment Currency | Commitment Frequency | 24-Sep    | 24-Aug | Total (last month excluded from total) |
| Duck, Donald and Daisy  |                   | 0                 | USD                 |                      | 333333333 | 0      | 333333333                              |
| Totals                  |                   |                   |                     |                      | 333333333 | 0      | 895                                    |
| Currency                | USD               | $                 |                     |                      |           |        |                                        |
| Partner                 | Status            | Commitment Amount | Commitment Currency | Commitment Frequency | 24-Sep    | 24-Aug | Total (last month excluded from total) |
| Duck, Daffy and Daphney | partner_financial | 30                | USD                 | Monthly              | 84        | 0      | 84                                     |
| Totals                  |                   |                   |                     |                      | 84        | 0      | 84                                     |

# React

| Currency                | USD               | $                 |                     |                      |           |        |                                        |
| ----------------------- | ----------------- | ----------------- | ------------------- | -------------------- | --------- | ------ | -------------------------------------- |
| Partner                 | Status            | Commitment Amount | Commitment Currency | Commitment Frequency | 24-Sep    | 24-Aug | Total (last month excluded from total) |
| Duck, Daffy and Daphney | partner_financial | 30                | USD                 | Monthly              | 84        | 0      | 84                                     |
| Totals                  |                   |                   |                     |                      | 84        | 0      | 84                                     |
| Currency                | UGS               | UGS               |                     |                      |           |        |                                        |
| Partner                 | Status            | Commitment Amount | Commitment Currency | Commitment Frequency | 24-Sep    | 24-Aug | Total (last month excluded from total) |
| Duck, Donald and Daisy  |                   | 0                 | USD                 |                      | 333333333 | 0      | 333333333                              |
| Totals                  |                   |                   |                     |                      | 333333333 | 0      | 333333333                              |
| Currency                | GBP               | £                 |                     |                      |           |        |                                        |
| Partner                 | Status            | Commitment Amount | Commitment Currency | Commitment Frequency | 24-Sep    | 24-Aug | Total (last month excluded from total) |
| Duck, Daffy and Daphney | partner_financial | 30                | USD                 | Monthly              | 0         | 10000  | 10000                                  |
| Totals                  |                   |                   |                     |                      | 0         | 10000  | 10000                                  |

canac commented on src/components/Reports/FourteenMonthReports/FourteenMonthReport.tsx:
```diff
@@ -109,148 +100,128 @@ export const FourteenMonthReport: React.FC<Props> = ({
-              )
-            : '';
+          const inHandMonthlyEquivalent =
+            contact.status === 'Partner - Financial' &&
```
I think it could be a future improvement if you're OK with that since this isn't new code, just code moved into a separate file. Since this data comes from the REST proxy, it isn't being translated into a ContactStatusEnum automatically for us. It won't break when task phases goes live because that status wasn't renamed, right?

wrandall22 replied:
That is correct. We could put a TODO in here to refactor later to use the id instead of the display value.

canac replied:
Actually, the contact status did change in https://github.com/CruGlobal/mpdx_api/pull/2799, so I'll have to fix that. I think I'll do it in a PR based on this one because this one is just supposed to be about creating one table per currency.

canac replied:
#1080 should fix this

canac commented:
@wrandall22 Do you mind reviewing this since Daniel is going to be out for the rest of the week?

wrandall22 commented on src/components/Reports/FourteenMonthReports/Layout/Table/helpers.test.ts:
```diff
@@ -0,0 +1,101 @@
+        makeContact(400, 250, 125),
+      ]),
+    ).toEqual([
+      { month: 'Jan', total: 570 },
+      { month: 'Feb', total: 540 },
+      { month: 'Mar', total: 315 },
```
I can't quite tell where these totals are coming from. I see `450`, `200`, and `775`.

wrandall22 replied:
Nvm, figured it out.

wrandall22 commented on pages/api/Schema/reports/fourteenMonth/fourteenMonth.graphql:
```diff
@@ -41,7 +41,7 @@ type FourteenMonthReportContact {
   average: Float!
   minimum: Float!
-  months: [FourteenMonthReportContactMonths!]
+  months: [FourteenMonthReportContactMonths!]!
```
Does the API give us back an empty array or something, guaranteeing never to be null or missing from the response?

canac replied:
As far as I can tell, `summarize_months` has to return an array because if `group_donations_by_month` doesn't return an array, the `.map` would throw https://github.com/CruGlobal/mpdx_api/blob/f1f4f35b42bef9483e1eff0d3c8f4003bc111437/app/services/reports/donor_currency_donations.rb#L58-L69

For what it's worth, on the JS side, if `months` ever were an array, this line would throw too. The proxy endpoint couldn't actually handle null months before.
https://github.com/CruGlobal/mpdx-react/blob/b35116164f7a6b9ba6cdcb6b75a9f5d75eb9ee0f/pages/api/Schema/reports/fourteenMonth/datahandler.ts#L114
